lib/fsm/logger.go

```diff
@@ -38,35 +40,68 @@ type Logger struct {
 	Operator ops.Operator
 	// Server is the optional server that will be attached to log entries
 	Server *storage.Server
+
+	// logEntryC is used to queue log entries and unblock execution while etcd is down during upgrades
+	// this has the potential to lose queued log messages if the process dies while etcd is down
+	logEntryC chan ops.LogEntry
+
+	// logEntryOnce bootstraps the LogEntry queue on the first log entry
+	logEntryOnce sync.Once
+}
+
+func (l *Logger) initQueue() {
+	l.logEntryOnce.Do(func() {
+		// initialize the queue to a reasonably large value, to queue all the messages during etcd upgrade
+		l.logEntryC = make(chan ops.LogEntry, 4096)
+		go l.runQueue()
+	})
+}
+
+func (l *Logger) runQueue() {
+	for {
+		msg := <-l.logEntryC
```

Comment:
Well, looking at it, the fsm itself to my knowledge never goes out of scope for a process it's loaded within, as it's either loaded within a shell, or within a temporary agent process for the purposes of running that process and exiting. So the goroutine stops when the process exits.

Reply:
I suppose what I could do is just move it to a package singleton, just get's created once per process and lives until the process exits

Reply:
This will accumulate goroutines for phases that are executed locally. I agree that the installer is relatively short-lived so it might justify the simplicity.